designs/SearchCommandSpec.md

```diff
@@ -0,0 +1,106 @@
+
+## Who are the customers
+
+This targets all NuGet.exe and .NET users that want to be able to search for packages from the command line, rather than using NuGet Gallery or VS PMUI. Also, while `nuget.exe list` allows users to search different package sources for packages with a search query, the results are returned sorted by the package Id, rather than by relevance, as proposed for this Search command. Further, the Search command would provide more information on the packages than List currently does.
+
+## Goals
+
+* Providing search functionality for NuGet packages through the commandline for NuGet.exe
+* Implementing a prototype of this feature for dotnet.exe
+* Displaying the search results to users in a way that maximizes the visibility of the most relevant results
+
+## Solution
+
+The search feature leverages the search service provided by the NuGet API. Search targets multiple sources and displays all of their results, sorted by relevance, one source after another. The command line interface is simple: __./NuGet.exe search \<query\>__.
+
+```
+PS C:\> ./NuGet.exe search logging
+```
+
+Each search result displays the name, version, number of downloads, and a preview of the description of the package.
+
+```
+====================
+Source: NuGet.org
+--------------------
+> Microsoft.Extensions.Logging.Abstractions | 5.0.0-preview.6.20305.6 | Downloads: 345,145,935
+  Logging abstractions for Microsoft.Extensions.Logging.
+
+Commonly Used Types:
+Microsoft.Extensions.Lo...
+
+--------------------
+> Microsoft.Extensions.Logging | 5.0.0-preview.6.20305.6 | Downloads: 243,186,566
+  Logging infrastructure default implementation for Microsoft.Extensions.Logging.
+When using NuGet 3....
+
+--------------------
+> Microsoft.IdentityModel.Logging | 6.6.0 | Downloads: 120,177,247
+  Includes Event Source based logging support.
+
+--------------------
+> Microsoft.Extensions.Logging.Configuration | 5.0.0-preview.6.20305.6 | Downloads: 63,205,816
+```
+
+If a particular source returns no results for a query, it will display output that indicates this.
+
+```
+====================
+Source: dotnet-msbuild
+--------------------
+No results found.
+```
+
+### Optional Arguments
+
+Search provides users with some optional arguments that they can use to customize their search queries:
+
+| Name | Description | Usage |
+| ---  |     ---     |  :-:  |
+| PackageTypes | The package type to use to filter packages (if supported by the source). If the provided package type is not a valid package type as defined by the [Package Type document](https://github.com/NuGet/Home/wiki/Package-Type-%5BPacking%5D), an empty result will returned. | -PackageType `<Package Type>`|
```

Comment:
nit, maybe a better user experience is saying "package type filtering is not supported on this feed" instead of the same message that occurs if they have too specific of parameters and no results are found. Said another way, this problem of no package type filtering cannot be resolved by the user but no results probably can.

Reply:
I think we should drop PackageTypes from the initial implementation. 

1. It increases the scope of the feature.
1. We don't have super well understood use cases for package types yet. `dotnet tool search` is an obvious example but that has it's use right now. 

I'd prefer to add that at a later point based on community feedback.